rosbag2/src/rosbag2/writer.cpp

```diff
@@ -90,7 +111,52 @@ void Writer::write(std::shared_ptr<SerializedBagMessage> message)
     throw std::runtime_error("Bag is not open. Call open() before writing.");
   }
 
+  // Update the message count for the Topic.
+  ++topics_[message->topic_name].message_count;
+  ++message_count_;
+
+  start_time_ = std::min(start_time_, message->time_stamp);
+  end_time_ = std::max(end_time_, message->time_stamp);
+
   storage_->write(converter_ ? converter_->convert(message) : message);
 }
 
+bool Writer::should_split_bagfile() const
+{
+  if (max_bagfile_size_ == rosbag2_storage::storage_interfaces::MAX_BAGFILE_SIZE_NO_SPLIT) {
+    return false;
+  } else {
+    return storage_->get_bagfile_size() > max_bagfile_size_;
+  }
+}
+
+rosbag2_storage::BagMetadata Writer::generate_metadata_() const
```

Comment:
I don't feel super strong about it, but I think most of the items (if not all) of the metadata information could be updated in place without the need of additional class members such as `start_time_`, .. `topics_`. The `generate_metadata` function would then actually just translate into a `get_metadata` function.

Reply:
I spent some time attempting to embed `BagMetadata` into `Writer`. `BagMetadata` has `std::chrono::time_point<std::chrono::high_resolution_clock> starting_time` and `std::chrono::nanoseconds duration`. While it can be done, it requires converting between time formats at every message write. 

I found it cleaner to instead resole the time values when metadata is written out.

Reply:
I can understand that. If it makes things easier, we can change/enhance the metadata information with a `end_time` if that makes things easier.

Reply:
I think it would be best to do that in a separate PR and iterate on what we currently have.

How about this:
* I build a local copy of `BagMetadata` in `Writer`
* I also store a `end_time` variable in `Writer` and translate that into `duration` on metadata write in the dtor
* once all bagsplitting is in, I open an issue to move `end_time` from `Writer` into `BagMetadata`